### PR TITLE
feat: Cluster resources ADR

### LINKED ISSUE #18: ADR for cluster scoped resources

### PR DESCRIPTION
**Pull request by tumido**
Resolves: https://github.com/operate-first/blueprint/issues/18, https://github.com/operate-first/apps/issues/115
Part of: https://github.com/operate-first/support/issues/39

A draft implementing described changes is located here: https://github.com/operate-first/apps/pull/120 or check it out at [`tumido:cleanup`](https://github.com/tumido/apps/tree/cleanup)

### CODE REVIEW

File: docs/adr/0009-cluster-resources.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+- Status: accepted
+- Deciders: 4n4nd, anishasthana, HumairAK, tumido
+- Date: 2021-01-18
+
+Technical Story: [issue-1](https://github.com/operate-first/odh-moc-support/issues/39), [issue-2](https://github.com/open-infrastructure-labs/ops-issues/issues/2), [pr-2](https://github.com/operate-first/apps/pull/104)
+
+## Context and Problem Statement
+
+A full Git-Ops driven approach we decided to take in the previous ADRs, makes us face additional challenges when deploying cluster scoped resources. Resources like `Namespace`, `Group` etc. should be also defined declaratively. They should be deployable via ArgoCD and allowing users to use optional encryption layer for sensitive properties/objects that may be considered a PII risk (see [`ADR 0008`](./)). Additionally the resulting layout should be flexible enough that migrating a project between clusters and/or extending the project to a multiple clusters is simple enough.
```
Comment (HumairAK) [inline docs/adr/0009-cluster-resources.md]: Rephrase: 
> They should be deployable via ArgoCD and allowing users to use optional encryption layer for sensitive 

to: 

> They should be deployable via ArgoCD and allow users to use an optional encryption layer for sensitive

File: docs/adr/0009-cluster-resources.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+- Deployable via ArgoCD with cluster admin permissions.
+- Common layout that can be adopted by various downstreams and target cluster deployments.
+- Provides consistent secret management.
+- Simplifies application onboarding to ArgoCD driven deployment as a single PR.
+- Can support cluster scoped resources of various kind as well as namespaced scoped resources that are sensitive and tied to the deployed cluster scoped - e.g. a `Namespace` resource and `admin` role assignment via a `RoleBiniding`s.
+
+## Considered Options
+
+- A `cluster-scope` application deployed via ArgoCD.
+-
```
Comment (HumairAK) [inline docs/adr/0009-cluster-resources.md]: missing bullet point?

File: docs/adr/0009-cluster-resources.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+A full Git-Ops driven approach we decided to take in the previous ADRs, makes us face additional challenges when deploying cluster scoped resources. Resources like `Namespace`, `Group` etc. should be also defined declaratively. They should be deployable via ArgoCD and allowing users to use optional encryption layer for sensitive properties/objects that may be considered a PII risk (see [`ADR 0008`](./)). Additionally the resulting layout should be flexible enough that migrating a project between clusters and/or extending the project to a multiple clusters is simple enough.
+
+## Decision Drivers
+
+- Declarative manifests in git repo.
+- Deployable via ArgoCD with cluster admin permissions.
+- Common layout that can be adopted by various downstreams and target cluster deployments.
+- Provides consistent secret management.
+- Simplifies application onboarding to ArgoCD driven deployment as a single PR.
+- Can support cluster scoped resources of various kind as well as namespaced scoped resources that are sensitive and tied to the deployed cluster scoped - e.g. a `Namespace` resource and `admin` role assignment via a `RoleBiniding`s.
```
Comment (HumairAK) [inline docs/adr/0009-cluster-resources.md]: I _think_ I understand what you're trying to say but I'm still not sure. Are you saying the following: 

> We should be able to deploy cluster scoped resources along with any supplementary namespace scoped resources, with the added ability to encrypt these resources as necessary.

**Comment by HumairAK** [comment]
@tumido 

What do you think of tying in operators, clusterrolebindings, etc. into the language here along with namespaces/groups/etc. As these will be manifests we commonly have to deploy.

[Relevant](https://github.com/open-infrastructure-labs/ops-issues/issues/6)

File: docs/adr/0009-cluster-resources.md (hunk 1 above)
Comment (anishasthana) [reply docs/adr/0009-cluster-resources.md]: Nit: I'd also remove `,` from `take in the previous ADRs, makes us face additional challenges`

Comment (anishasthana) [reply docs/adr/0009-cluster-resources.md]: `extending the project to a multiple clusters` -> `extending the project to multiple clusters`

File: docs/adr/0009-cluster-resources.md (hunk 3 above)
Comment (anishasthana) [reply docs/adr/0009-cluster-resources.md]: I think we should rephrase this point, I'm certainly very confused.

**Review by anishasthana — CHANGES_REQUESTED** [review]

**Comment by tumido** [comment]
> What do you think of tying in operators, clusterrolebindings, etc

@HumairAK very good point! I'll include that. :slightly_smiling_face:

File: docs/adr/0009-cluster-resources.md (hunk 1 above)
Comment (tumido) [reply docs/adr/0009-cluster-resources.md]: Thank you, both! :slightly_smiling_face:

File: docs/adr/0009-cluster-resources.md (hunk 3 above)
Comment (tumido) [reply docs/adr/0009-cluster-resources.md]: Rephrased:
> Allows to collocate cluster scoped resources along with any supplementary namespace scoped resources.

The encryption stuff was added to the 4th point as:

> Allows to collocate cluster scoped resources along with any supplementary namespace scoped resources.

File: docs/adr/0009-cluster-resources.md (hunk 2 above)
Comment (tumido) [reply docs/adr/0009-cluster-resources.md]: yes.. I wanted to add the current `apps` layout and the lars repo with groups and stuff, and forgot about it. Will update.

**Comment by 4n4nd** [comment]
@tumido I think this is great :+1:

**Review by 4n4nd — APPROVED** [review]

File: docs/adr/0009-cluster-resources.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        ...
+        └── bar                        # Example namespace overlayed from base
+            └── kustomization.yaml
+```
+
+Each namespace shared between clusters is expected to be hosted in a folder within `base`. All supplementary namespaced resources which requires cluster admin permissions are collocated here as well. Example: `RoleBinding`s for `ClusterRoles`, `Subscription`s for operators etc. If a namespace is environment (cluster) specific, it is expected to live within environment the overlay only.
+
+Disadvantages:
+
+- Too verbose since it requires many files and folders to be created when accomodatin minor changes
```
Comment (anishasthana) [inline docs/adr/0009-cluster-resources.md]: /s/accomodatin/accomodating

**Review by anishasthana — CHANGES_REQUESTED** [review]

**Comment by anishasthana** [comment]
One minor mistake, but outside of that I think this looks good.

File: docs/adr/0009-cluster-resources.md (hunk 4 above)
Comment (tumido) [reply docs/adr/0009-cluster-resources.md]: resolved! :slightly_smiling_face:

**Review by anishasthana — APPROVED** [review]

**Pull request merged**